### PR TITLE
Allow to modify Bcache devices

### PR DESCRIPTION
**Pull request by joseivanlopez**
PBI: https://trello.com/c/cQo16PQ5/672-editing-an-in-memory-bcache-device-devices-and-cache-mode

### CODE REVIEW

File: storage/Devices/Bcache.cc [hunk 1]
```diff
@@ -125,9 +125,16 @@ namespace storage
 
 
     void
-    Bcache::attach_bcache_cset(BcacheCset* bcache_cset)
+    Bcache::add_bcache_cset(BcacheCset* bcache_cset)
```
Comment (aschnell) [inline storage/Devices/Bcache.cc]: What is the motivation for this change? The term "attach" is used at may places in this concern, e.g. by the bcache tool (as seen in Bcache::Impl::do_attach_bcache_cset()).

File: storage/Devices/BcacheImpl.cc [hunk 2]
```diff
@@ -407,6 +407,21 @@ namespace storage
     }
 
 
+    void
+    Bcache::Impl::remove_bcache_cset()
+    {
+	if(get_type() == BcacheType::FLASH_ONLY)
+	    ST_THROW(LogicException("A Caching Set cannot be detached from a Flash-only Bcache"));
```
Comment (aschnell) [inline storage/Devices/BcacheImpl.cc]: Here the function is called "remove_..." but the error message says "detach". Please be consistent with the terms (I suggest to follow the bcache project).

File: storage/Devices/Bcache.cc (hunk 1 above)
Comment (joseivanlopez) [reply storage/Devices/Bcache.cc]: As discussed on IRC, this was changed for consistency with the rest of the public API.

File: storage/Devices/BcacheImpl.cc (hunk 2 above)
Comment (joseivanlopez) [reply storage/Devices/BcacheImpl.cc]: I will adapt the messages.

File: storage/Devices/Bcache.h [hunk 3]
```diff
@@ -101,13 +101,22 @@ namespace storage
...
 	/**
-	 * Attach a caching set to the Bcache.
+	 * Associates a caching set to the Bcache.
 	 *
 	 * This method does not make sense for Flash-only Bcache devices.
 	 *
 	 * @throw LogicException, Exception
 	 */
-	void attach_bcache_cset(BcacheCset* bcache_cset);
+	void add_bcache_cset(BcacheCset* bcache_cset);
```
Comment (aschnell) [inline storage/Devices/Bcache.h]: I assume the ABI break here is considered to be OK.

File: storage/Devices/BcacheImpl.cc [hunk 4]
```diff
@@ -576,6 +591,42 @@ namespace storage
     }
 
 
+    void
+    Bcache::Impl::add_modify_actions(Actiongraph::Impl& actiongraph, const Device* lhs) const
+    {
+	const Bcache* lhs_bcache = to_bcache(lhs);
+
+	// TODO Flash-only Bache
```
Comment (aschnell) [inline storage/Devices/BcacheImpl.cc]: typo "bache"

File: storage/Devices/BcacheImpl.cc [hunk 5]
```diff
@@ -725,6 +776,117 @@ namespace storage
...
+	return sformat(text, toString(get_cache_mode()), get_name(), get_size_text());
+    }
+
+
+    void
+    Bcache::Impl::do_update_cache_mode() const
+    {
+	string cmd_line = BCACHE_BIN " set-cachemode " + quote(get_backing_device()->get_name()) + " " + quote(toString(get_cache_mode()));
+
+	SystemCmd cmd2(cmd_line, SystemCmd::DoThrow);
```
Comment (aschnell) [inline storage/Devices/BcacheImpl.cc]: The "2" in "cmd2" is confusing.

**Comment by aschnell** [comment]
Only some minor stuff, LGTM.

File: storage/Devices/Bcache.h (hunk 3 above)
Comment (joseivanlopez) [reply storage/Devices/Bcache.h]: Yes, because `attach_bcache_cset` was added during SP1 development phase. So we are still in time to break it.
